File: app/src/main/java/com/simon/harmonichackernews/utils/ReferenceLinkRowUtils.java
package com.simon.harmonichackernews.utils;

import android.text.TextUtils;

public final class ReferenceLinkRowUtils {

    private ReferenceLinkRowUtils() {
    }

    public static String getReferenceLinkLabel(CollectedReferenceLinks.ReferenceLink link) {
        String resolvedTitle = link.getResolvedTitle();
        if (!TextUtils.isEmpty(resolvedTitle)) {
            return resolvedTitle.replace('\n', ' ').replaceAll("\\s+", " ").trim();
        }

        String label = link.getLabel();
        if (TextUtils.isEmpty(label)) {
            return link.getUrl();
        }
        return label.replace('\n', ' ').replaceAll("\\s+", " ").trim();
    }
}

File: app/src/main/java/com/simon/harmonichackernews/utils/ReferenceLinkRowUtils.kt
package com.simon.harmonichackernews.utils

import com.simon.harmonichackernews.utils.CollectedReferenceLinks.ReferenceLink

object ReferenceLinkRowUtils {
    private val whitespace = "\\s+".toRegex()

    fun getReferenceLinkLabel(link: ReferenceLink): String {
        val label = link.resolvedTitle?.takeIf(String::isNotEmpty)
            ?: link.label?.takeIf(String::isNotEmpty)
            ?: return link.url.orEmpty()
        return label.replace('\n', ' ').replace(whitespace, " ").trim { it <= ' ' }
    }
}
